contracts/builder-subnets/BuilderSubnets.sol

```diff
@@ -0,0 +1,502 @@
+        if (isMigrationOver) {
+            require(stakerAddress_ == _msgSender(), "BS: invalid sender");
+        }
+
+        BuildersSubnet storage subnet = buildersSubnets[subnetId_];
+        require(amount_ > 0, "BS: nothing to stake");
+        require(block.timestamp >= subnet.startsAt, "BS: stake isn't started");
+
+        Staker storage staker = stakers[subnetId_][stakerAddress_];
+
+        uint256 staked_ = staker.staked + amount_;
+        require(staked_ >= subnet.minStake, "BS: staked amount too low");
+
+        IERC20(token).safeTransferFrom(_msgSender(), address(this), amount_);
+
+        claimLockEnd_ = uint128(
+            claimLockEnd_.max(subnet.minClaimLockEnd).max(staker.claimLockEnd).max(block.timestamp)
+        );
+
+        _updateStorage(subnetId_, stakerAddress_, staked_, claimLockEnd_);
+        staker.lastStake = uint128(block.timestamp);
+
+        emit Staked(subnetId_, stakerAddress_, staker);
+    }
+
+    function withdraw(bytes32 subnetId_, uint256 amount_) external onlyExistedSubnet(subnetId_) {
+        address stakerAddress_ = _msgSender();
+
+        BuildersSubnet storage subnet = buildersSubnets[subnetId_];
+        Staker storage staker = stakers[subnetId_][stakerAddress_];
+        if (amount_ > staker.staked) {
+            amount_ = staker.staked;
+        }
+        require(amount_ > 0, "BS: nothing to withdraw");
+        uint256 minAllowedWithdrawalTimestamp = staker.lastStake + subnet.withdrawLockPeriodAfterStake;
+        require(block.timestamp > minAllowedWithdrawalTimestamp, "BS: user withdraw is locked");
+        uint256 staked_ = staker.staked - amount_;
+        require(staked_ >= subnet.minStake || staked_ == 0, "BS: min stake reached");
+
+        _updateStorage(subnetId_, stakerAddress_, staked_, staker.claimLockEnd);
+
+        (uint256 fee_, address treasuryAddress_) = _getProtocolFee(amount_, FEE_WITHDRAW_OPERATION);
+        if (fee_ > 0) {
+            IERC20(token).safeTransfer(treasuryAddress_, fee_);
+            amount_ -= fee_;
+
+            emit FeePaid(subnetId_, stakerAddress_, fee_, treasuryAddress_);
+        }
+
+        IERC20(token).safeTransfer(stakerAddress_, amount_);
+
+        emit Withdrawn(subnetId_, stakerAddress_, staker, amount_);
+    }
+
+    function claim(bytes32 subnetId_, address stakerAddress_) external onlyExistedSubnet(subnetId_) {
+        Staker storage staker = stakers[subnetId_][stakerAddress_];
+        require(block.timestamp > staker.claimLockEnd, "BS: claim is locked");
+
+        _updateStorage(subnetId_, stakerAddress_, staker.staked, staker.claimLockEnd);
+        uint256 toClaim_ = staker.pendingRewards;
+        staker.pendingRewards = 0;
+
+        require(toClaim_ > 0, "BS: nothing to claim");
+
+        (uint256 protocolFee_, address protocolTreasury_) = _getProtocolFee(toClaim_, FEE_CLAIM_OPERATION);
+        (uint256 subnetFee_, address subnetTreasury_) = _getSubnetFee(toClaim_, subnetId_);
```

Comment:
We can calculate a subnetFee based on (toClaim_ - protocolFee_). What is better?